ibmcloud: Added test cases to check peer pod with large image

sudharshanibm3 opened the pull request:
Added test cases to check peer pod pulls larger image in ibmcloud cloud-provider

Fixes: #1111

stevenhorsman commented:
Hey @sudharshanibm3 - can you list the output that you get as it's difficult to test before the image has been built. Thanks!

sudharshanibm3 commented:
Output log for `TestCreatePeerPodWithLargeImage` :
```
=== RUN   TestCreatePeerPodWithLargeImage
=== RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test
=== RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created
    common_suite_test.go:175: Pod with larger image started successfully
        ImageName:quay.io/sudharshanibm3/test-images:largerpod
        TimeTaken:406.823292ms
time="2023-06-22T23:04:31+05:30" level=info msg="PodVM name: largeimagepod"
time="2023-06-22T23:04:33+05:30" level=debug msg="Instance number: 0, Instance id: 0787_0e4065cc-9ac6-454e-962f-084b1e0cadf4, Instance name: podvm-largeimagepod-2d13fe23"
time="2023-06-22T23:04:33+05:30" level=info msg="Deleting pod... largeimagepod"
--- PASS: TestCreatePeerPodWithLargeImage (58.16s)
    --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test (58.16s)
        --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created (2.60s)
PASS
ok      github.com/confidential-containers/cloud-api-adaptor/test/e2e339.082s
```

stevenhorsman commented:
I'm not sure I understand the output.
```
TimeTaken:406.823292ms
```
We are trying to time the remote pod pull, which I'd expect to take much longer than that for a large image. Ideally we were looking to test that a pod that took > 60s to pull didn't time out.

sudharshanibm3 commented:
Hi @stevenhorsman,
 - Used following docker file to create a large image
 ```
FROM --platform="${TARGETPLATFORM}" alpine:3.17
RUN dd if=/dev/random of=rand-data bs=4GB count=1
```

- Built and pushed image to `quay.io/sudharshanibm3/test-images:largeimage`
- Following are the events of the pod
```
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  22s   default-scheduler  Successfully assigned default/largeimagepod to 10.242.0.174
  Normal  Pulling    1s    kubelet            Pulling image "quay.io/sudharshanibm3/test-images:largeimage"
  Normal  Pulled     1s    kubelet            Successfully pulled image "quay.io/sudharshanibm3/test-images:largeimage" in 414.181167ms (416.296933ms including waiting)
  Normal  Created    1s    kubelet            Created container largeimagepod
sudharshanmuralidharan@Sudharshans-MacBook-Pro cloud-api-adaptor % 
```

stevenhorsman commented:
> * quay.io/sudharshanibm3/test-images:largeimage

Is that the events of it being pulled locally as it could be cached. What does the cloud-api-adaptor log say about the pull image timing?

stevenhorsman commented:
For reference the docker pull on my VM took 2m54, so I think the image should be big enough:
```
$ time docker pull quay.io/sudharshanibm3/test-images:largeimage
largeimage: Pulling from sudharshanibm3/test-images
4db1b89c0bd1: Pull complete
acf00e44ce0a: Pull complete
Digest: sha256:5bb44ee8440609bc930a50b793301d90395201e85515bd8f3bf5684f288f9b6a
Status: Downloaded newer image for quay.io/sudharshanibm3/test-images:largeimage
quay.io/sudharshanibm3/test-images:largeimage

real	2m54.213s
user	0m0.320s
sys	0m0.385s
```

sudharshanibm3 commented:
> Is that the events of it being pulled locally as it could be cached. What does the cloud-api-adaptor log say about the pull image timing?

CAA log:

```
2023/06/22 18:26:55 [adaptor/proxy] getImageName: got image from annotations: quay.io/sudharshanibm3/test-images:largeimage
2023/06/22 18:26:55 [adaptor/proxy] CreateContainer: calling PullImage for "quay.io/sudharshanibm3/test-images:largeimage" before CreateContainer (cid: "8783c169f42fc6946d45c51d8f7b7f371ae3c909faf33367c44710a6fdff43c2")
2023/06/22 18:27:28 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:largeimage"
2023/06/22 18:27:36 [adaptor/proxy] StartContainer: containerID:8783c169f42fc6946d45c51d8f7b7f371ae3c909faf33367c44710a6fdff43c2
2023/06/22 18:27:46 [adaptor/proxy] RemoveContainer: containerID:8783c169f42fc6946d45c51d8f7b7f371ae3c909faf33367c44710a6fdff43c2
```

stevenhorsman commented:
> > Is that the events of it being pulled locally as it could be cached. What does the cloud-api-adaptor log say about the pull image timing?
> 
> CAA log:
> 
> ```
> 2023/06/22 18:26:55 [adaptor/proxy] getImageName: got image from annotations: quay.io/sudharshanibm3/test-images:largeimage
> 2023/06/22 18:26:55 [adaptor/proxy] CreateContainer: calling PullImage for "quay.io/sudharshanibm3/test-images:largeimage" before CreateContainer (cid: "8783c169f42fc6946d45c51d8f7b7f371ae3c909faf33367c44710a6fdff43c2")
> 2023/06/22 18:27:28 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:largeimage"
> 2023/06/22 18:27:36 [adaptor/proxy] StartContainer: containerID:8783c169f42fc6946d45c51d8f7b7f371ae3c909faf33367c44710a6fdff43c2
> 2023/06/22 18:27:46 [adaptor/proxy] RemoveContainer: containerID:8783c169f42fc6946d45c51d8f7b7f371ae3c909faf33367c44710a6fdff43c2
> ```

Ok, that suggests that the pull took ~33s rather than the milliseconds, so I think we need to update the logic being used to calculate the pulling time.

sudharshanibm3 commented:
Hi @stevenhorsman ,
1. Updated code with new logic for calculating Pulling time of an image by using `startTime` on Pulling and `endTime` on Started status
2. Calculated the time differences by `endTime.Sub(startTime)`
3. Now the time taken on pulling the image is approximately `40.769794036s`
Could you please let me know your views on this new logic for calculating pulling time of an image?

Output Log:
```
=== RUN   TestCreatePeerPodWithLargeImage
=== RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test
=== RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created
    common_suite_test.go:176: Pod with larger image started successfully
        ImageName:quay.io/sudharshanibm3/test-images:largeimage
        TimeTaken:40.769794036s
time="2023-06-25T23:50:09+05:30" level=info msg="Deleting pod... largeimage105pod"
--- PASS: TestCreatePeerPodWithLargeImage (60.84s)
    --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test (60.84s)
        --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created (55.21s)
PASS
ok      github.com/confidential-containers/cloud-api-adaptor
```

stevenhorsman commented:
> Hi @stevenhorsman ,
> 
> 1. Updated code with new logic for calculating Pulling time of an image by using `startTime` on Pulling and `endTime` on Started status
> 2. Calculated the time differences by `endTime.Sub(startTime)`
> 3. Now the time taken on pulling the image is approximately `40.769794036s`
>    Could you please let me know your views on this new logic for calculating pulling time of an image?
> 
> Output Log:
> 
> ```
> === RUN   TestCreatePeerPodWithLargeImage
> === RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test
> === RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created
>     common_suite_test.go:176: Pod with larger image started successfully
>         ImageName:quay.io/sudharshanibm3/test-images:largeimage
>         TimeTaken:40.769794036s
> time="2023-06-25T23:50:09+05:30" level=info msg="Deleting pod... largeimage105pod"
> --- PASS: TestCreatePeerPodWithLargeImage (60.84s)
>     --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test (60.84s)
>         --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created (55.21s)
> PASS
> ok      github.com/confidential-containers/cloud-api-adaptor
> ```

Yeah - this logic looks much better thanks! I'm wondering if we should consider making the test image 8gb instead so that the image pull goes over the 60s timeout issues we've seen in the past if this is reliably taken around 40s at the moment?

sudharshanibm3 commented:
Hi @stevenhorsman ,
- Built 2 images with sizes [6GB](quay.io/sudharshanibm3/test-images:testlargeimage6gb) and [8GB](quay.io/sudharshanibm3/test-images:testlargeimage8gb)
- Created a new cluster and tried running test cases with above images (to avoid container cache)
- Image with 6GB took about `2m20.959208516s` to pull the image

```
=== RUN   TestCreatePeerPodWithLargeImage6gb
=== RUN   TestCreatePeerPodWithLargeImage6gb/LargeImagePeerPod6gb_test
=== RUN   TestCreatePeerPodWithLargeImage6gb/LargeImagePeerPod6gb_test/Peer_pod_with_Large_Image_6gb_has_been_created
    common_suite_test.go:176: Pod with larger image started successfully
        ImageName:quay.io/sudharshanibm3/test-images:testlargeimage6gb
        TimeTaken:2m20.959208516s
time="2023-06-26T16:38:56+05:30" level=info msg="Deleting pod... largeimage156pod-6gb"
--- PASS: TestCreatePeerPodWithLargeImage6gb (164.31s)
    --- PASS: TestCreatePeerPodWithLargeImage6gb/LargeImagePeerPod6gb_test (164.31s)
        --- PASS: TestCreatePeerPodWithLargeImage6gb/LargeImagePeerPod6gb_test/Peer_pod_with_Large_Image_6gb_has_been_created (158.73s)
```


- Whereas image with 8GB Pulled the image but Failed to create container

```
=== RUN   TestCreatePeerPodWithLargeImage8gb
=== RUN   TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test
=== RUN   TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test/Peer_pod_with_Large_Image_8gb_has_been_created
    common_suite_test.go:174: Error: failed to create containerd task: failed to create shim task: context deadline exceeded
time="2023-06-26T16:42:38+05:30" level=info msg="Deleting pod... largeimage25pod-8gb"
--- FAIL: TestCreatePeerPodWithLargeImage8gb (222.06s)
    --- FAIL: TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test (222.06s)
        --- FAIL: TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test/Peer_pod_with_Large_Image_8gb_has_been_created (216.47s)
FAIL
```

Following are the pod events of image with 8GB

```
Events:
  Type     Reason     Age    From               Message
  ----     ------     ----   ----               -------
  Normal   Scheduled  3m43s  default-scheduler  Successfully assigned default/largeimage25pod-8gb to 10.242.0.167
  Normal   Pulling    3m18s  kubelet            Pulling image "quay.io/sudharshanibm3/test-images:testlargeimage8gb"
  Normal   Pulled     66s    kubelet            Successfully pulled image "quay.io/sudharshanibm3/test-images:testlargeimage8gb" in 2m11.500439219s (2m11.500455348s including waiting)
  Normal   Created    63s    kubelet            Created container largeimage25pod-8gb
  Warning  Failed     2s     kubelet            Error: failed to create containerd task: failed to create shim task: context deadline exceeded
```

stevenhorsman commented:
ok, that's good information, can you show the logs of the CAA for the 8GB pull so we can have an idea what timeout might are hitting here?

It's odd that the 4GB image took 40s earlier and now 130s. Any ideas of why that might be?

stevenhorsman commented:
That suggests that in both scenarios the pull took around 23s. What did you test report as you had figures of ~40s/1m40s I thought, does that mean that you timing logic is still out?

sudharshanibm3 commented:
Hi @stevenhorsman ,
- Updated code with new logic
- Written code to calculate image pull-time directly from CAA pod Logs

Test output log:
```
=== RUN   TestCreatePeerPodWithLargeImage
=== RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test
=== RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created
2023/06/26 19:43:58 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:largeimage" end....
    common_suite_test.go:179: Pod with larger image started successfully
    common_suite_test.go:180: ImageName:quay.io/sudharshanibm3/test-images:largeimage
    common_suite_test.go:181: TimeTaken:25s
time="2023-06-27T01:14:12+05:30" level=info msg="PodVM name: largeimage-27-pod"
time="2023-06-27T01:14:14+05:30" level=debug msg="Instance number: 0, Instance id: 0787_e0101f35-c4d3-4174-9dce-c02175e357e7, Instance name: podvm-largeimage-27-pod-7013f69f"
time="2023-06-27T01:14:14+05:30" level=info msg="Deleting pod... largeimage-27-pod"
--- PASS: TestCreatePeerPodWithLargeImage (63.80s)
    --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test (63.80s)
        --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created (3.23s)
PASS
```


CAA Log :

```
2023/06/26 19:43:33 [adaptor/proxy] getImageName: got image from annotations: quay.io/sudharshanibm3/test-images:largeimage
2023/06/26 19:43:33 [adaptor/proxy] CreateContainer: calling PullImage for "quay.io/sudharshanibm3/test-images:largeimage" before CreateContainer (cid: "7ca7c376358cdec13730a2c75a113579062aeaa295a1ae705cb83ebd45e2f9d3")
2023/06/26 19:43:58 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:largeimage"
2023/06/26 19:44:09 [adaptor/proxy] StartContainer: containerID:7ca7c376358cdec13730a2c75a113579062aeaa295a1ae705cb83ebd45e2f9d3
2023/06/26 19:44:15 [adaptor/proxy] RemoveContainer: containerID:7ca7c376358cdec13730a2c75a113579062aeaa295a1ae705cb83ebd45e2f9d3
2023/06/26 19:44:28 [adaptor/proxy] RemoveContainer fails: rpc error: code = Internal desc = destroy cgroups


```

stevenhorsman commented:
Great - thanks. Can you now the run the tests with 6GB and 8GB containers so we can try and work out if/where the timeout issue is?

sudharshanibm3 commented:
Test with 6GB image:
- Timetaken: `31s`
- Test logs: 
```
=== RUN   TestCreatePeerPodWithLargeImage6gb
=== RUN   TestCreatePeerPodWithLargeImage6gb/LargeImagePeerPod6gb_test
=== RUN   TestCreatePeerPodWithLargeImage6gb/LargeImagePeerPod6gb_test/Peer_pod_with_Large_Image_6gb_has_been_created
2023/06/27 08:50:11 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:testlargeimage6gb" end....
    common_suite_test.go:179: Pod with larger image started successfully
    common_suite_test.go:180: ImageName:quay.io/sudharshanibm3/test-images:testlargeimage6gb
    common_suite_test.go:181: TimeTaken:31s
time="2023-06-27T14:20:24+05:30" level=info msg="PodVM name: largeimage-45-pod-6gb"
time="2023-06-27T14:20:25+05:30" level=debug msg="Instance number: 0, Instance id: 0787_a75e3579-7bdc-44dd-a418-34a12d7d4e4d, Instance name: podvm-largeimage-45-pod-6gb-f26e030f"
time="2023-06-27T14:20:25+05:30" level=info msg="Deleting pod... largeimage-45-pod-6gb"
--- PASS: TestCreatePeerPodWithLargeImage6gb (159.35s)
    --- PASS: TestCreatePeerPodWithLargeImage6gb/LargeImagePeerPod6gb_test (159.35s)
        --- PASS: TestCreatePeerPodWithLargeImage6gb/LargeImagePeerPod6gb_test/Peer_pod_with_Large_Image_6gb_has_been_created (3.80s)
```
- CAA Logs:
```
2023/06/27 08:49:40 [adaptor/proxy] getImageName: got image from annotations: quay.io/sudharshanibm3/test-images:testlargeimage6gb
2023/06/27 08:49:40 [adaptor/proxy] CreateContainer: calling PullImage for "quay.io/sudharshanibm3/test-images:testlargeimage6gb" before CreateContainer (cid: "367610d10b2d51c2d065e374d5a43238484fca7ced3abbaa5efc64609ddac040")
2023/06/27 08:50:11 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:testlargeimage6gb"
2023/06/27 08:50:16 [adaptor/proxy] StartContainer: containerID:367610d10b2d51c2d065e374d5a43238484fca7ced3abbaa5efc64609ddac040
2023/06/27 08:50:26 [adaptor/proxy] RemoveContainer: containerID:367610d10b2d51c2d065e374d5a43238484fca7ced3abbaa5efc64609ddac040
2023/06/27 08:50:28 [podnetwork] routes on netns /var/run/netns/cni-ffa68550-ed41-1ac1-240c-cbc30ddfe195
```


Test with 8GB Image:
- Failed to Start Container
- Test Logs:
```
=== RUN   TestCreatePeerPodWithLargeImage8gb
=== RUN   TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test
E0627 14:24:41.072660    8170 request.go:1058] Unexpected error when reading response body: http2: client connection lost
    common_suite_test.go:134: unexpected error when reading response body. Please retry. Original error: http2: client connection lost
--- FAIL: TestCreatePeerPodWithLargeImage8gb (255.38s)
    --- FAIL: TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test (255.38s)
FAIL
FAIL    github.com/confidential-containers/cloud-api-adaptor/test/e2e   545.150s
FAIL
```
- CAA Logs:
```
2023/06/27 08:54:40 [adaptor/proxy] getImageName: got image from annotations: quay.io/sudharshanibm3/test-images:testlargeimage8gb
2023/06/27 08:54:40 [adaptor/proxy] CreateContainer: calling PullImage for "quay.io/sudharshanibm3/test-images:testlargeimage8gb" before CreateContainer (cid: "159e277c981a677b6a532658309073672cf7ad2307b80393ae31f745db00fca8")
2023/06/27 08:55:40 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:testlargeimage8gb"
2023/06/27 08:55:40 [adaptor/proxy] CreateContainer fails: context deadline exceeded
2023/06/27 08:55:44 [adaptor/proxy] RemoveContainer: containerID:7b4eaf2d441392da87fdfa7df7391fb9a7151dc38176625d5438297bdcab7d92
2023/06/27 08:55:52 [adaptor/proxy] RemoveContainer fails: rpc error: code = Internal desc = destroy cgroups
```

stevenhorsman commented:
So my interpretation of those results is that we still have the 60sec timeout issue, and that 8GB was just about the image size that triggered that (for this container registry and network settings), so we arguable fail the test of the timeout. This is good info to understand though.

sudharshanibm3 commented:
Sorry for the confusion @stevenhorsman , previous test might failed may be because of network issue

Ran the test cases another time
- Now it pulled the 8GB image
- within `38s`
- Test log:
```
=== RUN   TestCreatePeerPodWithLargeImage8gb
=== RUN   TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test
=== RUN   TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test/Peer_pod_with_Large_Image_8gb_has_been_created
2023/06/27 09:28:00 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:testlargeimage8gb" end....
    common_suite_test.go:179: Pod with larger image started successfully
    common_suite_test.go:180: ImageName:quay.io/sudharshanibm3/test-images:testlargeimage8gb
    common_suite_test.go:181: TimeTaken:38s
time="2023-06-27T14:58:14+05:30" level=info msg="PodVM name: largeimage-756-pod-8gb"
time="2023-06-27T14:58:15+05:30" level=debug msg="Instance number: 0, Instance id: 0787_80ff706a-96f3-49ea-af5c-b5fe13304c4a, Instance name: podvm-largeimage-756-pod-8gb-2a8efe06"
time="2023-06-27T14:58:15+05:30" level=info msg="Deleting pod... largeimage-756-pod-8gb"
--- PASS: TestCreatePeerPodWithLargeImage8gb (78.55s)
    --- PASS: TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test (78.55s)
        --- PASS: TestCreatePeerPodWithLargeImage8gb/LargeImagePeerPod8gb_test/Peer_pod_with_Large_Image_8gb_has_been_created (2.93s)
PASS
ok      github.com/confidential-containers/cloud-api-adaptor/test/e2e   247.555s
```

- CAA Log:
```
2023/06/27 09:27:22 [adaptor/proxy] getImageName: got image from annotations: quay.io/sudharshanibm3/test-images:testlargeimage8gb
2023/06/27 09:27:22 [adaptor/proxy] CreateContainer: calling PullImage for "quay.io/sudharshanibm3/test-images:testlargeimage8gb" before CreateContainer (cid: "3ad6f4060ccb21a220d46e51bdfbf826fb015dde9e542f9c57da47cc12d6fc0e")
2023/06/27 09:28:00 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:testlargeimage8gb"
2023/06/27 09:28:08 [adaptor/proxy] StartContainer: containerID:3ad6f4060ccb21a220d46e51bdfbf826fb015dde9e542f9c57da47cc12d6fc0e
2023/06/27 09:28:16 [adaptor/proxy] RemoveContainer: containerID:3ad6f4060ccb21a220d46e51bdfbf826fb015dde9e542f9c57da47cc12d6fc0e
2023/06/27 09:28:28 [adaptor/proxy] RemoveContainer fails: rpc error: code = Internal desc = destroy cgroups
```

stevenhorsman commented:
Ok, that's also interesting to know. Maybe the answer is that we pick a pod size e.g. 4GB as a sensible 'max supported' and have the tests for that and not worry too much about the timeout if that reliably works?

sudharshanibm3 commented:
Hi @stevenhorsman ,
- Updated code and Dockerfile for building image with size 4GB

Test log:
```
=== RUN   TestCreatePeerPodWithLargeImage
=== RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test
=== RUN   TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created
2023/06/27 13:03:45 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:largeimage" end....
    common_suite_test.go:179: Pod with larger image started successfully
    common_suite_test.go:180: ImageName:quay.io/sudharshanibm3/test-images:largeimage
    common_suite_test.go:181: TimeTaken:25s
time="2023-06-27T18:34:03+05:30" level=info msg="PodVM name: largeimage-920-pod"
time="2023-06-27T18:34:06+05:30" level=debug msg="Instance number: 0, Instance id: 0787_6ac3fcad-739a-443a-b5fd-24b84b172640, Instance name: podvm-largeimage-920-pod-dcc2712b"
time="2023-06-27T18:34:06+05:30" level=info msg="Deleting pod... largeimage-920-pod"
--- PASS: TestCreatePeerPodWithLargeImage (75.22s)
    --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test (75.22s)
        --- PASS: TestCreatePeerPodWithLargeImage/LargeImagePeerPod_test/Peer_pod_with_Large_Image_has_been_created (4.61s)
PASS
ok      github.com/confidential-containers/cloud-api-adaptor/test/e2e   135.284s
```


CAA Log:
```
2023/06/27 13:03:20 [adaptor/proxy] getImageName: got image from annotations: quay.io/sudharshanibm3/test-images:largeimage
2023/06/27 13:03:20 [adaptor/proxy] CreateContainer: calling PullImage for "quay.io/sudharshanibm3/test-images:largeimage" before CreateContainer (cid: "61da300568af7233bc01807c48ed2146461ddece0afc294d8876f2d9735db89f")
2023/06/27 13:03:45 [adaptor/proxy] CreateContainer: successfully pulled image "quay.io/sudharshanibm3/test-images:largeimage"
2023/06/27 13:03:57 [adaptor/proxy] StartContainer: containerID:61da300568af7233bc01807c48ed2146461ddece0afc294d8876f2d9735db89f
2023/06/27 13:04:07 [adaptor/proxy] RemoveContainer: containerID:61da300568af7233bc01807c48ed2146461ddece0afc294d8876f2d9735db89f
2023/06/27 13:04:16 [adaptor/proxy] RemoveContainer fails: rpc error: code = Internal desc = destroy cgroups
```

stevenhorsman commented on test/e2e/common_suite_test.go:
```diff
@@ -52,6 +54,7 @@ type testCase struct {
 	testCommands         []testCommand
 	expectedPodLogString string
 	podState             v1.PodPhase
+	podWatcher           bool
```
I wonder if imagePullTimer would be a better name for it based on the current implementation, or if you want a generically named `podWatcher` then move the image pull logging into `watchImagePullTime` and then pass it as a function parameter?

stevenhorsman commented on test/e2e/common_suite_test.go:
```diff
@@ -254,11 +281,71 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
+	}
+	return slice
+}
+func watchImagePullTime(ctx context.Context, client klient.Client, caaPod v1.Pod, Pod v1.Pod) (string, error) {
```
new line the separate the functions please

stevenhorsman commented on test/e2e/common_suite_test.go:
```diff
@@ -254,11 +281,71 @@ func newTestCase(t *testing.T, testName string, assert CloudAssert, assessMessag
+			podLogSlice := reverseSlice(strings.Split(podLogString, "\n"))
+			for _, i := range podLogSlice {
+				if strings.Contains(i, "calling PullImage for \""+Pod.Spec.Containers[0].Image+"\"") {
+					startTime, err = time.Parse("15:04:05", strings.Split(i, " ")[1])
```
I wonder if using regex, or `scan_fmt` for this matching and extraction the time string might cleaner?

stevenhorsman commented on test/e2e/common_suite_test.go:
```diff
@@ -157,8 +165,24 @@ func (tc *testCase) run() {
+							if err != nil {
+								t.Fatal(err)
+							}
+							t.Logf("Pod with larger image started successfully")
+							t.Logf("ImageName:%s", tc.pod.Spec.Containers[0].Image)
+							t.Logf("TimeTaken:%s", imagePullTime)
```
I wonder if this logging could be better than:
```
    common_suite_test.go:179: Pod with larger image started successfully
    common_suite_test.go:180: ImageName:quay.io/sudharshanibm3/test-images:largeimage
    common_suite_test.go:181: TimeTaken:25s
```
Maybe it could go into a single log line and be more like Time taken to pull 4GB image: ?

sudharshanibm3 commented:
Hi @stevenhorsman 
1. Created a [timeExtractor](https://github.com/sudharshanibm3/cloud-api-adaptor/blob/e2e-test-larger-pod-image/test/e2e/common_suite_test.go#L298-L304) function for comparing and extracting time from Log String using regex
2. Renamed `podWatcher ` to `watchImagePullTime `
3. Converted output as single Log

mattarnoatibm commented on test/e2e/common_suite_test.go:
```diff
@@ -249,16 +275,93 @@ func (tc *testCase) run() {
+				}
+			}
+		} else {
+			return "", errors.New("Pod Failed to Start")
```
I think we need a different error message string here since in this case the pod is in the Running state but somehow has produced no pod logs.

mattarnoatibm commented on test/e2e/common_suite_test.go:
```diff
@@ -521,3 +624,11 @@ func doTestCreatePeerPodAndCheckEnvVariableLogsWithImageAndDeployment(t *testing
+
+func doTestCreatePeerPodWithLargeImage(t *testing.T, assert CloudAssert) {
+	namespace := envconf.RandomName("default", 7)
+	podName := "largeimage-" + strconv.Itoa(rand.Intn(1000)) + "-pod"
```
Why does the pod name of the large image pod contain a random substring?

sudharshanibm3 replied:
Hi @mattarnoatibm ,
Initially we assessed the pod completion status based on event logs of pod. In order to avoid repeated events during assessment, we used to generate pod with different name each time. Now we changed our approach in assessment by checking the Logs of CAA. Will remove the random substring from podName now. Thanks!

mattarnoatibm commented on test/e2e/common_suite_test.go:
```diff
@@ -249,16 +275,93 @@ func (tc *testCase) run() {
 	return testCase
 }
 
+func reverseSlice(slice []string) []string {
```
Can we use the `slices.Reverse` function rather than write a custom function to reverse s string slice https://pkg.go.dev/slices@master#Reverse?

sudharshanibm3 replied:
```
test/e2e/common_suite_test.go:13:2: package slices is not in GOROOT (/usr/local/go/src/slices)
FAIL    github.com/confidential-containers/cloud-api-adaptor/test/e2e [setup failed]
FAIL
make: *** [test-e2e] Error 1
```

stevenhorsman replied:
I think slices.Reverse might be go 1.21+ which we aren't on yet?

sudharshanibm3 commented:
Hi @mattarnoatibm 
Updated code with requested changes

mattarnoatibm reviewed (approved):
LGTM

stevenhorsman reviewed (approved):
LGTM. Thanks
